### PR TITLE
4. Hyper-parameter tuning (extension with XGBoost)

### PR DESCRIPTION
**Pull request by jvmead**
an extension to the ML lessons, picking up where k-fold left off
currently runs into problems with the CI timing out, presumably due to long running time?
```
https://colab.research.google.com/
Notebook real time --- 1051.062778711319 seconds ---
Notebook CPU time --- 728.3288249999999 seconds ---
```
The successfully executed notebook with plots generated using full stats may be found here:
https://colab.research.google.com/drive/1I34R8eCck3wo1YX54WllUKJw9KdoID8Y

### CODE REVIEW

File: advanced-python/4bHyperparameterTuning.ipynb [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+    "import uproot\n",
+    "\n",
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "import xgboost as xgb\n",
+    "\n",
+    "from xgboost.sklearn import XGBClassifier\n",
+    "from sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier\n",
+    "\n",
+    "from sklearn import cross_validation, metrics\n",
```
Comment (chrisburr) [inline advanced-python/4bHyperparameterTuning.ipynb]: I think you can fix the test using this as `cross_validation` isn't used?
```suggestion
    "from sklearn import metrics\n",
```

**Comment by chrisburr** [comment]
It looks like there are some more imports to fix

File: advanced-python/4bHyperparameterTuning.ipynb [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#max_entries = 1000\n",
+    "data_df = uproot.open('/eos/user/l/lhcbsk/advanced-python/data/real_data.root')['DecayTree'].pandas.df()#entrystop=max_entries)\n",
```
Comment (chrisburr) [inline advanced-python/4bHyperparameterTuning.ipynb]: We shouldn't rely on the EOS mount. Those files are available with HTTPS:
* `https://cern.ch/starterkit/data/advanced-python-2018/real_data.root`
* `https://cern.ch/starterkit/data/advanced-python-2018/simulated_data.root`

Comment (jvmead) [reply advanced-python/4bHyperparameterTuning.ipynb]: ah okay, I'd just thought I'd forgotten 'root://eosuser.cern.ch/', thanks!

Comment (chrisburr) [reply advanced-python/4bHyperparameterTuning.ipynb]: I'd rather use `root://` to be more realistic but unfortunately it's not possible to make parts of `root://eosuser.cern.ch/` completely public.

**Comment by jvmead** [comment]
@chrisburr down to <20mins on colab, is there any leeway we can get on the CI?

**Comment by jvmead** [comment]
@chrisburr unsure of what the issue is, and how to proceed with the last step:
```test $TRAVIS_PULL_REQUEST == "false" && test $TRAVIS_BRANCH == "master" && starterkit_ci deploy```

**Comment by chrisburr** [comment]
You can ignore that, it’s the bit that uploads it to the website but we can fix that later

**Comment by jvmead** [comment]
@chrisburr do I need permission to access ```starterkit_ci```?
https://github.com/jvmead/analysis-essentials/runs/476855433
```
/home/runner/work/analysis-essentials/analysis-essentials/advanced-python/4bModelTuning.ipynb:document isn't included in any toctree
Traceback (most recent call last):
  File "/usr/share/miniconda/envs/my-analysis-env/bin/starterkit_ci", line 8, in <module>
    sys.exit(parse_args())
  File "/usr/share/miniconda/envs/my-analysis-env/lib/python3.6/site-packages/starterkit_ci/__init__.py", line 35, in parse_args
    allow_warnings=args.allow_warnings
  File "/usr/share/miniconda/envs/my-analysis-env/lib/python3.6/site-packages/starterkit_ci/__init__.py", line 44, in build_docs
    _sphinx_build('html', source_dir, allow_warnings)
  File "/usr/share/miniconda/envs/my-analysis-env/lib/python3.6/site-packages/starterkit_ci/__init__.py", line 82, in _sphinx_build
    return check_call(cmd, cwd=source_dir)
  File "/usr/share/miniconda/envs/my-analysis-env/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['sphinx-build', '-M', 'html', '.', 'build', '-W']' returned non-zero exit status 2.
##[error]Process completed with exit code 1.
```

**Comment by jvmead** [comment]
I had a look in the [starterkit-ci](https://github.com/lhcb/starterkit-ci ) but couldn't see where this [problem ](https://github.com/jvmead/analysis-essentials/runs/476855433) arises

**Comment by chrisburr** [comment]
You need to add it to one of the tables of contents. The error is complaining that the page is unreachable from any of the menus.

**Comment by jvmead** [comment]
nice, thanks

**Comment by jvmead** [comment]
created an issue: https://github.com/hsf-training/analysis-essentials/issues/25
to view the lesson being added: https://colab.research.google.com/drive/13dwXKKHxqQfk8Zo2Gzh46_pCc9xBgi1H?usp=sharing

**Comment by jonas-eschle** [comment]
Hi, what's the status of this? The build seems to fail

**Comment by jvmead** [comment]
As far as I remember, the issue was that the CI would time out

I think this was an issue with the notebook taking 20-30mins to run
Chris also suspected issues with its inclusion in the table of contents tree

The notebooks may now also be out of sync with the current SK build as this was almost 2 years ago now :)

**Comment by jonas-eschle** [comment]
Okay, so since the build time takes up already quite a lot, it would be great oc to move this build time down somehow. What about using a smaller sample? 10%, 5% of the sample? (and note that this is done for demonstration purposes?

I would just suggest to either give this a little push now or close as nomerge (the longer it's here, the lower the odds that it gets merged at some point). What do you think?

**Comment by jvmead** [comment]
I've got entrystop set to 1000 events and am just waiting for the dependencies to finish installing (currently at ~1h30)

**Comment by jvmead** [comment]
Install dependancies:
```
6h 0m 8s
Run source ${CONDA}/etc/profile.d/conda.sh
Collecting package metadata (repodata.json): ...working... done
Error: The operation was canceled.
```
https://github.com/jvmead/analysis-essentials/runs/4422343032?check_suite_focus=true

**Comment by jonas-eschle** [comment]
The branch uses (an old?) a version of the github workflow which is different from the master (and uses stuff such as the `source conda`)
The file seems to have gotten renamed to build instead of CI. Can you maybe rebase or merge master into your branch and push again?

**Comment by jvmead** [comment]
@jonas-eschle I'm a little lost at this point
[https://github.com/jvmead/analysis-essentials/runs/5050083345?check_suite_focus=true#step:4:116](https://github.com/jvmead/analysis-essentials/runs/5050083345?check_suite_focus=true#step:4:116)
I'm not sure if this is a change in uproot or the dataset

**Comment by jonas-eschle** [comment]
This is indeed a change in uproot (in 4), an example of how to get the same now is using can be found here: https://github.com/hsf-training/analysis-essentials/blob/master/advanced-python/20DataAndPlotting.ipynb (with the .array you can convert it to np or pd for pandas)

**Comment by jvmead** [comment]
Fantastic, thanks very much!

**Comment by jvmead** [comment]
@jonas-eschle now I'm seeing problems with advanced-python/20DataAndPlotting.ipynb
https://github.com/jvmead/analysis-essentials/runs/5052417479?check_suite_focus=true#step:4:57
I've not made any changes to that file since pulling from master

**Comment by jonas-eschle** [comment]
Hm, that should be only temporary, a problem with retrieving the files over the network

**Comment by jonas-eschle** [comment]
Can you maybe restart the CI? I think this was a temporary problem that was solved now

**Comment by jvmead** [comment]
I've pushed a trivial change, just updating the link to the version on google collab which contains plots for comparison.
The version on collab also has a couple of cells added at the end which I think might be worth making note of in the lesson:
https://colab.research.google.com/drive/1I34R8eCck3wo1YX54WllUKJw9KdoID8Y#scrollTo=8Xlc0Wik0xfY 
https://colab.research.google.com/drive/1I34R8eCck3wo1YX54WllUKJw9KdoID8Y#scrollTo=q6hRFVQHTXQ-&line=2&uniqifier=1

**Comment by jonas-eschle** [comment]
now it looks fine! However, the document is not yet included anywhere in the toctree. Maybe it could be as an additional tutorial, like an advanced? Or with the normal ones? What do you suggest?

**Comment by jvmead** [comment]
When I demonstrated in the advanced python session, the notebook for lesson 4 had a title for k-folding but no content beyond that. This notebook was meant to pick up where that one left off so there probably isn't time to allow students to run everything in the class but it might be worth having the notebook available on the site while the lecturer can scroll through the pre-run more complete version (https://bit.ly/LHCb_XGB_Tuning) to show the plots and discuss the process and allow people to un-comment the extra dimensions of the hyperparameter scan and test it in their own time.

**Comment by klieret** [comment]
Hi @jonas-eschle @jvmead It seemed to me that you were almost ready to merge this, right? What's still blocking this?

**Comment by jvmead** [comment]
I wasn't sure how to resolve this and have been away from my desk for the better part of a month
https://github.com/hsf-training/analysis-essentials/runs/6352614562?check_suite_focus=true#step:4:1255

**Comment by jonas-eschle** [comment]
Yes, they should be uploaded without an image, but executing them should produce the image. So I am not sure I understand the problem. Currently not really available, but otherwise ping me again later to look into this if it still persists

**Comment by jvmead** [comment]
I haven't touched any notebook but my own so I'm not sure where this issue is arising, thanks I'd appreciate if you could take a look upon your return

**Comment by jvmead** [comment]
@jonas-eschle any chance you could take another look at this?

**Comment by jonas-eschle** [comment]
I will next week, thanks for the ping

**Comment by jonas-eschle** [comment]
@jvmead finally unstalled it, the CI works again etc, back to this!
Why is the ci.yaml needed? We do already have the building?

And can we rename it to 33 maybe? So that all the 3 notebooks are classification/BDT/ML?

File: advanced-python/10Basics.ipynb
Comment (jonas-eschle) [inline advanced-python/10Basics.ipynb:1]: Not sure about this change, is this needed?

**Review by jonas-eschle — CHANGES_REQUESTED** [review]
Also, the ci.yaml should then be removed, right?

**Comment by jonas-eschle** [comment]
For the additions, I would suggest to get this merged, and then quickly make a new PR with the additions?
I think it just needs to be cleaned up a bit (see also review comments) but other than that, it looks good to me

(sidenote, for future reference: name your branch more meaningful, i.e. paramtune, bdttutorial_hyper or whatever, not simply Patch-1 (which is a github default, I am aware))

File: advanced-python/10Basics.ipynb
Comment (jvmead) [reply advanced-python/10Basics.ipynb:1]: yes, I'm assuming those were mistakenly committed

**Comment by jonas-eschle** [comment]
Okay, sounds good. Can you remove the files and undo the changes?

>  The [`.github` directory it resides in](https://github.com/jvmead/analysis-essentials/blob/patch-1/.github/workflows/ci.yaml) doesn't even exist on main.

Not quite, it does, but not the fork you created (maybe use the "sync fork" button in the repo).

Btw, if you encounter technical difficulties along the way now, let me know (and if it's easier to add the changes now, we can too)

**Comment by jvmead** [comment]
my bad! I didn't mean to hit ready for review yet, I'm just reverting the changes to lesson 10 and 11 now

**Comment by jonas-eschle** [comment]
@jvmead , there seems to be an issue in the notebook, can you check and verify that you can run it locally?

**Comment by jvmead** [comment]
Should be all good now!

**Comment by klieret** [comment]
Congrats on merging this big PR!  Btw, what the bot just commented is right: We'd be happy to add you to the community page @jvmead and @jonas-eschle :)

**Comment by jvmead** [comment]
Hi @klieret,
I'd be happy to be listed! I've added an .md file and made a merge request if that's still all that's necessary. I did notice just now that this lesson's inclusion in the advanced python contents page isn't quite right. Can this be edited trivially or should another fork be made?

**Pull request merged**